delta-lake/delta-spark332db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuMergeIntoCommand.scala

```diff
@@ -0,0 +1,1171 @@
+        createMetric(sc, "number of target partitions after skipping"),
+    "numTargetPartitionsRemovedFrom" ->
+        createMetric(sc, "number of target partitions from which files were removed"),
+    "numTargetPartitionsAddedTo" ->
+        createMetric(sc, "number of target partitions to which files were added"),
+    "executionTimeMs" ->
+        createMetric(sc, "time taken to execute the entire operation"),
+    "scanTimeMs" ->
+        createMetric(sc, "time taken to scan the files for matches"),
+    "rewriteTimeMs" ->
+        createMetric(sc, "time taken to rewrite the matched files"))
+
+  override def run(spark: SparkSession): Seq[Row] = {
+    recordDeltaOperation(targetDeltaLog, "delta.dml.merge") {
+      val startTime = System.nanoTime()
+      gpuDeltaLog.withNewTransaction { deltaTxn =>
+        if (target.schema.size != deltaTxn.metadata.schema.size) {
+          throw DeltaErrors.schemaChangedSinceAnalysis(
+            atAnalysis = target.schema, latestSchema = deltaTxn.metadata.schema)
+        }
+
+        if (canMergeSchema) {
+          updateMetadata(
+            spark, deltaTxn, migratedSchema.getOrElse(target.schema),
+            deltaTxn.metadata.partitionColumns, deltaTxn.metadata.configuration,
+            isOverwriteMode = false, rearrangeOnly = false)
+        }
+
+        val deltaActions = {
+          if (isSingleInsertOnly && spark.conf.get(DeltaSQLConf.MERGE_INSERT_ONLY_ENABLED)) {
+            writeInsertsOnlyWhenNoMatchedClauses(spark, deltaTxn)
+          } else {
+            val filesToRewrite = findTouchedFiles(spark, deltaTxn)
+            val newWrittenFiles = withStatusCode("DELTA", "Writing merged data") {
+              writeAllChanges(spark, deltaTxn, filesToRewrite)
+            }
+            filesToRewrite.map(_.remove) ++ newWrittenFiles
+          }
+        }
+
+        // Metrics should be recorded before commit (where they are written to delta logs).
+        metrics("executionTimeMs").set((System.nanoTime() - startTime) / 1000 / 1000)
+        deltaTxn.registerSQLMetrics(spark, metrics)
+
+        // This is a best-effort sanity check.
+        if (metrics("numSourceRowsInSecondScan").value >= 0 &&
+            metrics("numSourceRows").value != metrics("numSourceRowsInSecondScan").value) {
+          log.warn(s"Merge source has ${metrics("numSourceRows").value} rows in initial scan but " +
+              s"${metrics("numSourceRowsInSecondScan").value} rows in second scan")
+          if (conf.getConf(DeltaSQLConf.MERGE_FAIL_IF_SOURCE_CHANGED)) {
+            throw DeltaErrors.sourceNotDeterministicInMergeException(spark)
+          }
+        }
+
+        deltaTxn.commit(
+          deltaActions,
+          DeltaOperations.Merge(
+            Option(condition.sql),
+            matchedClauses.map(DeltaOperations.MergePredicate(_)),
+            notMatchedClauses.map(DeltaOperations.MergePredicate(_)),
+            // TODO do we need to pass something in here?
```

Comment:
Yes, there needs to be support added here (along with integration tests of that support).  It's likely to be similar to delta-io/delta#1511.  Might be best done as a separate PR.  Note that these clauses are likely passed into the CPU MergeIntoCommand[Edge] instance.  Could key off of that to fallback to the CPU if it's non-empty in the interim so we're at least doing something safe in that unsupported case.

Curious if the Delta Lake integration tests have been manually tested against this shim?  I would not expect premerge to do it, since premerge doesn't know about this shim yet and isn't setup to run on a Databricks 12.2 instance.

Reply:
> Curious if the Delta Lake integration tests have been manually tested against this shim? I would not expect premerge to do it, since premerge doesn't know about this shim yet and isn't setup to run on a Databricks 12.2 instance.

I've been manually running the integration tests, but now I realize that the delta lake tests were being skipped. There are some failures, and I am looking into those now.